account_financial_report_webkit/models/account_static_balance.py

```diff
@@ -53,119 +61,155 @@ def unlink(self):
-            per_ids, acc_ids = map(list, map(set, zip(*missing_combinations)))
+            per_ids, acc_ids, journal_ids = map(
+                list, map(set, zip(*missing_combinations)))
         missing_periods = self.env['account.period'].browse(per_ids)
         missing_accounts = self.env['account.account'].browse(acc_ids)
-        return missing_periods, missing_accounts
+        missing_journals = self.env['account.journal'].browse(journal_ids)
+        return missing_periods, missing_accounts, missing_journals
 
     @api.model
-    def calculate_static_balance(self, periods, accounts=None):
+    def calculate_static_balance(self, periods, accounts=None, journals=None):
         """ Calculate the balance for a specified set of periods
         and insert them into the table."""
         self = self.suspend_security()
-        periods = periods.filtered(
-            lambda p: p.state == 'done' and not p.special)
-        for row in self.calculate_balance(periods, accounts=accounts):
-            try:
-                with self.env.cr.savepoint():
-                    self.create(row)
-            except psycopg2.IntegrityError:
-                # A balance already exists for this period and account
-                pass
+
+        if not journals:
+            # When journals are set, we can assume that this is a close for
+            # a journal period. Otherwise we need to filter the periods to
+            # check if they are in a done state
+            periods = periods.filtered(
+                lambda p: p.state == 'done' and not p.special)
+
+        for period in periods:
+            for row in self.calculate_balance(
+                    period, accounts=accounts, journals=journals):
+                try:
+                    with self.env.cr.savepoint():
+                        self.create(row)
+                except psycopg2.IntegrityError:
+                    # A balance already exists for this period and account
+                    pass
 
     @api.model
-    def calculate_balance(self, periods, accounts=None, include_draft=False):
-        """Calculate the balance for a specified set of periods
-        :param periods: periods to calculate the values for
+    def calculate_balance(self, period, accounts=None, include_draft=False,
+                          journals=None):
+        """
+        Calculate the balance for a specified set of periods
+        :param period: period to calculate the values for
         :param accounts: optional accounts to calculate the values for
+        :param journals: optional journals to calculate the values for
         :return: True
         """
-        if not periods:
+        if not period:
+            return {}
+
+        # Implement fallback when accounts or journals are None
+        if not accounts or not journals:
+            move_lines = self.env['account.move.line'].search([
+                ('period_id', '=', period.id)
+            ])
```

Comment:
Won't this lead to a massive recordset? Wouldn't a query selecting individual accounts and journals be more efficient here?

Reply:
You mean two queries like:
```SELECT account_id FROM account_move_line WHERE period_id = %S GROUP BY account_id``` and the same for journal_id?

Reply:
Execution time of the query above is 5 minutes.
SELECT account_id FROM account_move_line WHERE period_id = %S GROUP BY account_id